lib/one_gadget/fetchers/base.rb

```diff
@@ -112,19 +113,80 @@ def resolve_execve_args(processor, arg0, arg1, arg2, allow_null_argv: true)
         { constraints: cons, envp: envp }
       end
 
-      # arg == NULL || [arg] == NULL
-      def check_execve_arg(processor, arg, allow_null)
-        if arg.start_with?(processor.sp) # arg = sp+<num>
-          # in this case, the only chance is [sp+<num>] == NULL
-          num = Integer(arg[processor.sp.size..-1])
-          slot = processor.stack[num].to_s
-          return if global_var?(slot)
+      def check_argv(processor, arg, allow_null)
+        lmda = OneGadget::Emulators::Lambda.parse(arg)
 
-          "#{slot} == NULL"
-        elsif allow_null
-          "[#{arg}] == NULL || #{arg} == NULL"
+        if lmda.deref_count.zero? && OneGadget::ABI.stack_register?(lmda.obj)
+          return check_stack_argv(processor, lmda, allow_null)
+        end
+
+        check_nonstack_argv(arg, allow_null)
+      end
+
+      def check_stack_argv(processor, lmda, allow_null)
+        stack = processor.get_corresponding_stack(lmda.obj)
+        argv = (0..3).map { |i| stack[lmda.immi + processor.class.bits / 8 * i].to_s }
+
+        # if argv is already valid, no constraints are needed! (but probably won't happen :p)
+        return if argv_already_valid?(argv)
+
+        if global_var?(argv[0])
+          # argv[0] is not controlled by the user, argv[0] probably is "/bin/sh" or "sh" (but actually, the content of
+          # argv[0] doesn't quite matter, just need to make sure it's readable)
+          # So far (I checked glibc 2.37), we can make argv to be {"/bin/sh", sth, NULL} or {"sh", "-c", sth, NULL}
+          # TODO: We need to update this when the above assumption is no longer true
```

Comment:
These descriptions would be more suitable to be moved to above `generate_argv_with_sh`